### DISCUSSION TITLE
Is it possible to integrate something similar to Mabinogi/Archeage/etc for custom music?

### DISCUSSION THREAD

**Discussion by dra6o0n** [Yellow-Dog-Man/Resonite-Issues#2105] [Questions & Answers] [answered]
Games like those are MMO and uses Music Macro Language to play back audio samples in real time on all clients, rather than a url to say a youtube link or a stream link for audio, this is more like real time midi playback with the 'instructions' either sent in real time to anyone in the area (this implies every client needs to have this embedded and active).

[MML example](https://www.youtube.com/watch?v=zGYF5xJLYl4)
Note how it parses texts in real time, this means if someone walks in while the middle of the music is played they will hear it from that point on.

In games like Final Fantasy 14, you can use a bard class and play instruments in real time as well as playing specific notes will actually sync across the server to anyone in the area that can see and hear it.

**Comment by Frooxius** [comment, marked as answer]
Technically it's possible, but this is usually more of a question - would we invest the time into it? 

This seems a bit too specific/niche for something like Resonite, so I don't think we'd integrate these specifically, unless there was really big amount of interest in these.

**Reply by dra6o0n** [reply]
Yeah, though custom music is one of the main reasons why some very old MMO like Mabinogi even stays relevant for it's 16th year, though Resonite pretty has that same element in the form of custom user created content through scripting and modelling objects and such.

The interest would have to come from people who likes to compose music ingame than from a external program.

**Reply by JackTheFoxOtter** [reply]
MML from my understanding is basically a human-readable string notation for MIDI music. It itself might be a bit nieche, but general MIDI support would be super valuable to allow people to compose music within Resonite.

I see a couple of different related general feature requests:
- Support for Standard MIDI File playback (so it spawns an audio interface when I import a .mid file)
- The ability to parse a SMF file to extract individual nodes, as #1598 already requests
- The ability to synthesize individual tones as they would appear in an SMF file. That would allow users to create their own string notation parsers (for example for MML). That said, I think this might make most sense to be an additional node for #567

**Reply by AmasterAmaster** [reply]
I would love to see MIDI support in Resonite. Not sure if it can be a separate feature request, because seeing Jack's response of other people wanting similar requests that pertain to it, but if it helps with getting that type of support, that would be awesome.

**Reply by dra6o0n** [reply]
> MML from my understanding is basically a human-readable string notation for MIDI music. It itself might be a bit nieche, but general MIDI support would be super valuable to allow people to compose music within Resonite.
> 
> I see a couple of different related general feature requests:
> 
>     * Support for Standard MIDI File playback (so it spawns an audio interface when I import a .mid file)
> 
>     * The ability to parse a SMF file to extract individual nodes, as [Support for reading/parsing MIDI files #1598](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1598) already requests
> 
>     * The ability to synthesize individual tones as they would appear in an SMF file. That would allow users to create their own string notation parsers (for example for MML). That said, I think this might make most sense to be an additional node for [Implement and integrate audio DSP system for ProtoFlux #567](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/567)

Parsing is very similar to it, but generally how client oriented games handle MML today is having Microsoft GS Wavetable Synth which exists on every Windows PC, to playback sounds... The thing here is that this synth is very basic compared to say certain games that prefers higher quality soundfonts in form of DLS files.

I use FL studio as a hobby to transcribe game music and I do see DLS files involved in a lot of midi and other synthesized notes and instruments. In the midi settings, the MS GS Wavetable Synth is listed there for playback basically.

> The Microsoft GS Wavetable Synth can be found on every Windows computer since 1998. It was the default MIDI playback engine, whose soundfont consisted of downgraded versions of various Roland Sound Canvas instruments. The SC55 piano was unmistakeable, and it still pops up in memes.

So basically for 'playback', this engine is the default engine that will function when midi is played. I'm not sure how this playback occurs in actual applications though.
